server/src/main/java/org/elasticsearch/common/Rounding.java

```diff
@@ -1132,50 +1132,69 @@ public long beforeOverlap(long localMillis, Overlap overlap) {
+                            final OffsetDateTime offsetTime = roundedLocalDateTime.atOffset(currentOffsets.get(offsetIndex));
+                            final Instant offsetInstant = offsetTime.toInstant();
+                            if (previousTransition != null && offsetInstant.isBefore(previousTransition.getInstant())) {
+                                /*
+                                 * Rounding down across the transition can yield the
+                                 * wrong result. It's best to return to the transition
+                                 * time and round that down.
+                                 */
+                                attempts++;
+                                utcMillis = previousTransition.getInstant().toEpochMilli() - 1;
+                                continue attempt;
+                            }
+
+                            if (utcInstant.isBefore(offsetTime.toInstant()) == false) {
+                                return offsetInstant.toEpochMilli();
+                            }
                         }
 
-                        if (utcInstant.isBefore(offsetTime.toInstant()) == false) {
-                            return offsetInstant.toEpochMilli();
-                        }
+                        final OffsetDateTime offsetTime = roundedLocalDateTime.atOffset(currentOffsets.get(0));
+                        final Instant offsetInstant = offsetTime.toInstant();
+                        throw new IllegalArgumentException(
+                            this + " failed to round " + utcMillis + " down: " + offsetInstant + " is the earliest possible"
+                        );
+                    } else {
+                        // The desired time isn't valid because within a gap, so just return the start of the gap
+                        ZoneOffsetTransition zoneOffsetTransition = timeZone.getRules().getTransition(roundedLocalDateTime);
+                        return zoneOffsetTransition.getInstant().toEpochMilli();
                     }
-
-                    final OffsetDateTime offsetTime = roundedLocalDateTime.atOffset(currentOffsets.get(0));
-                    final Instant offsetInstant = offsetTime.toInstant();
-                    assert false : this + " failed to round " + utcMillis + " down: " + offsetInstant + " is the earliest possible";
-                    return offsetInstant.toEpochMilli(); // TODO or throw something?
-                } else {
-                    // The desired time isn't valid because within a gap, so just return the start of the gap
-                    ZoneOffsetTransition zoneOffsetTransition = timeZone.getRules().getTransition(roundedLocalDateTime);
-                    return zoneOffsetTransition.getInstant().toEpochMilli();
                 }
+                throw new IllegalArgumentException(
+                    this
+                        + " failed to round "
```

Comment:
Can we please log the timezone and the interval in the exception message if this happens? This way if we have missed a case that we are not aware of we'll at least be able to debug it. Perhaps a logger warning too?

Reply:
I'll make sure it gets in there. I think folks intended `this.toString` to have the zone but I expect that got lost and the lack of unit test didn't notice.